---
title: "BL-CNV analysis, part IV"
author: "AG"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    df_print: paged
    code_folding: hide
    fig_width: 10
    fig_height: 6
    theme: cerulean
    highlight: kate
    toc_float: yes
  pdf_document:
    toc: yes
    toc_depth: 2
editor_options:
  markdown:
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

## Scheme of the region of interest

Compared to part 2, here I use longer both red and green regions. 
For the red one I will use all available length (1677 bp). 
The green region will have the same length.

![scheme](figures/pDA61218-116_long_flanking_regions.png)

## Overview of reads

replicate E

### Length distribution

```{bash, eval=F}
conda activate snakemake
seqkit watch resources/reads/replicate_e/CNV_reads_all.fastq.gz -O results/reads_stat/replicate_e.png -j 8
```

![histogram raw](../results/reads_stat/replicate_e.png)

Filter out all reads shorter than 3.7 Kb (FR+IS+FR) - minimal read length that can capture 0 CN variants consisting only of two FRs and one IS copy between them.

Precise length FR+IS+FR = 3756.6

```{bash, eval=FALSE}
# filter out too short reads
conda activate /home/andrei/mambaforge/envs/filtlong-env 
LENGTH="3757"
filtlong --min_length $LENGTH resources/reads/replicate_e/CNV_reads_all.fastq.gz > results/reads_filtered/replicate_e/CNV_reads_all_3.7k.fastq && pigz -p 12 results/reads_filtered/replicate_e/CNV_reads_all_3.7k.fastq
conda deactivate

# trim adapters
conda activate porechop-env
porechop -i results/reads_filtered/replicate_e/CNV_reads_all_3.7k.fastq.gz -o results/reads_filtered/replicate_e/CNV_reads_all_3.7k_chop.fastq.gz --threads 10
conda deactivate

# histogram
seqkit watch results/reads_filtered/replicate_e/CNV_reads_all_3.7k_chop.fastq.gz -O results/reads_stat/replicate_e_min_3.7k.png -j 12

# convert fastq to fasta
seqkit fq2fa results/reads_filtered/replicate_e/CNV_reads_all_3.7k_chop.fastq.gz > results/reads_filtered/replicate_e/CNV_reads_all_3.7k.fasta
```

![histogram 7k](../results/reads_stat/replicate_e_min_3.7k.png)

### Quality distribution

of reads longer than 3.7k

```{bash, eval=F}
seqkit watch results/reads_filtered/replicate_e/CNV_reads_all_3.7k_chop.fastq.gz -O results/reads_stat/replicate_e_qual_min3.7.png -j 8 -f MeanQual
```

![quality](../results/reads_stat/replicate_e_qual_min3.7.png)

## Blast the flankig regions 

left = red; right = green

```{bash, eval=F}
# RUN BLAST
#
# -sorthsps <Integer, (>=0 and =<4)>
#   Sorting option for hps:
#     0 = Sort by hsp evalue,
#     1 = Sort by hsp score,
#     2 = Sort by hsp query start,
#     3 = Sort by hsp percent identity,
#     4 = Sort by hsp subject start
#   Not applicable for outfmt != 0
#
# -num_alignments <Integer, >=0>
#   Number of database sequences to show alignments for
#   Default = `250'
#    * Incompatible with:  max_target_seqs

# num_alignments should be high enough, depending on the total number of reads
blastn -query resources/flanking_regions/flanking_region_red_1677.fa -subject results/reads_filtered/replicate_e/CNV_reads_all_3.7k.fasta -outfmt 6 -num_alignments 333000 -num_threads 4 > results/tables/blast_flanking_region_red.tsv

blastn -query resources/flanking_regions/flanking_region_green_1677.fa -subject results/reads_filtered/replicate_e/CNV_reads_all_3.7k.fasta -outfmt 6 -num_alignments 333000 -num_threads 4 > results/tables/blast_flanking_region_green.tsv
```

## Parse BLAST results

### Red

reads of interest should contain 'red' segment at their start

```{r, message=FALSE, warning=FALSE}
blast_red <- read_delim("../results/tables/blast_flanking_region_red.tsv", col_names = F)

names(blast_red) <- c("query", "subject", "identity", "length", "mismatch", 
                        "gaps", "start.query", "end.query", "start.subject", 
                        "end.subject", "e.value", "bit.score") 
blast_red <- 
  blast_red %>%
  mutate(orientation = if_else(start.subject < end.subject, "direct", "reverse"))

# rename query
blast_red$query <- "FR_red"

blast_red
```

#### Reads with multiple hits

```{r}
# rows with the same subject
blast_red %>% 
  group_by(subject) %>% 
  count() %>% 
  ggplot(aes(n)) + 
  geom_histogram(bins = 150, fill="darkred") +
  geom_rug()
```

```{r}
# rows with the same subject
blast_red %>% 
  group_by(subject, orientation) %>% 
  count() %>% 
  ggplot(aes(n)) + 
  geom_histogram(bins = 150, fill="darkred") +
  geom_rug() +
  facet_grid(rows = vars(orientation))
```

#### Hit length

```{r}
blast_red %>%
  ggplot(aes(length)) +
  geom_histogram(bins = 150, fill='darkred') 
```

Number of query hits shorter than 1500: `r blast_red %>% filter(length < 1500) %>% nrow()`

Number of query hits longer than 1500: `r blast_red %>% filter(length > 1500) %>% nrow()`

#### Hit identity

```{r}
blast_red %>%
  ggplot(aes(identity)) +
  geom_histogram(bins = 150, fill='darkred') 
```

Number of query hits with identity below 75%: `r blast_red %>% filter(identity < 75) %>% nrow()`

Number of query hits with identity above 75%: `r blast_red %>% filter(identity >= 75) %>% nrow()`

#### E-value

```{r}
blast_red %>%
  ggplot(aes(e.value)) +
  geom_histogram(bins = 150, fill='darkred') 
```

```{r}
blast_red %>%
  filter(e.value != 0) %>% 
  ggplot(aes(e.value)) +
  geom_histogram(bins = 150, fill='darkred') +
  ggtitle("e-value = 0 excluded")
```

Query hits with e-value above $ 10^-5$

```{r}
blast_red %>% 
  filter(e.value > 10**-5) %>% 
  nrow()
```

Also, these have very short lengths

```{r}
blast_red %>% 
  filter(e.value > 10**-5) %>% 
  select(length)
```

Remaining hits are reliable

#### Length vs Identity vs E-value

```{r}
blast_red %>% 
  ggplot(aes(length, identity)) +
  geom_point(aes(color=e.value)) +
  ggtitle("query hits raw")
```

Filtering criteria: 

- length > 1500
- e-value < 10**-5
- identity >= 75%

```{r}
blast_red %>% 
  filter(length > 1500,
         e.value < 10**-5,
         identity >= 75) %>% 
  ggplot(aes(length, identity)) +
  geom_point(aes(color=e.value)) +
  ggtitle("query hits filtered")
```

### Green

```{r, message=FALSE, warning=FALSE}
blast_green <- read_delim("../results/tables/blast_flanking_region_green.tsv", col_names = F)

names(blast_green) <- c("query", "subject", "identity", "length", "mismatch", 
                        "gaps", "start.query", "end.query", "start.subject", 
                        "end.subject", "e.value", "bit.score")
blast_green <- 
  blast_green %>%
  mutate(orientation = if_else(start.subject < end.subject, "direct", "reverse"))

blast_green$query <- "FR_green"

blast_green
```


#### Reads with multiple hits

```{r}
# rows with the same subject
blast_green %>% 
  group_by(subject) %>% 
  count() %>% 
  ggplot(aes(n)) + 
  geom_histogram(bins = 30, fill="darkgreen") +
  geom_rug() +
  xlab("n hits per read")
```

```{r}
# rows with the same subject
blast_green %>% 
  group_by(subject, orientation) %>% 
  count() %>% 
  ggplot(aes(n)) + 
  geom_histogram(bins = 30, fill="darkgreen") +
  geom_rug() +
  facet_grid(rows = vars(orientation))
```

#### Hit length

```{r}
blast_green %>%
  ggplot(aes(length)) +
  geom_histogram(bins = 150, fill='darkgreen') 
```

Number of query hits shorter than 1500: `r blast_green %>% filter(length < 1500) %>% nrow()`

Number of query hits longer than 1500: `r blast_green %>% filter(length > 1500) %>% nrow()`

```{r}
blast_green %>% filter(length < 1500)
```

#### Identity

```{r}
blast_green %>%
  ggplot(aes(identity)) +
  geom_histogram(bins = 150, fill='darkgreen') 
```

Number of query hits with identity below 75%: `r blast_green %>% filter(identity < 75) %>% nrow()`

Number of query hits with identity above 75%: `r blast_green %>% filter(identity >= 75) %>% nrow()`

#### E-value

```{r}
blast_green %>%
  ggplot(aes(e.value)) +
  geom_histogram(bins = 150, fill='darkgreen') 
```

```{r}
blast_green %>%
  filter(e.value != 0) %>% 
  ggplot(aes(e.value)) +
  geom_histogram(bins = 150, fill='darkgreen') +
  ggtitle("e-value = 0 excluded")
```

Query hits with e-value above $ 10^-5$

```{r}
blast_green %>% 
  filter(e.value > 10**-5) %>% 
  nrow()
```

Remaining hits are reliable

#### Length vs Identity vs E-value

```{r}
blast_green %>% 
  ggplot(aes(length, identity)) +
  geom_point(aes(color=e.value)) +
  ggtitle("query hits raw")
```

After filtering: 

- length > 1500
- e-value < 10**-5
- identity >= 75%

```{r}
blast_green %>% 
  filter(length > 1500,
         e.value < 10**-5,
         identity >= 75) %>% 
  ggplot(aes(length, identity)) +
  geom_point(aes(color=e.value)) +
  ggtitle("query hits filtered")
```

The plot looks very similar to the one with 'red' query hits

## Filtering, part 1

BLAST results filtering criteria for both *red* and *green* FRs

- length > 1500
- e-value < 10**-5
- identity >= 75%

Some reads have multiple hits (red and green)!

```{r}
blast_red_filt <-
  blast_red %>%
  filter(length > 1500,
         e.value < 10 ** -5,
         identity >= 75)

blast_green_filt <-
  blast_green %>%
  filter(length > 1500,
         e.value < 10 ** -5,
         identity >= 75)
```

### Multiple hits per read?

```{r}
blast_red_filt %>% 
  group_by(subject) %>% 
  count() %>% 
  ggplot(aes(n)) + 
  geom_histogram(bins = 150, fill="darkred") +
  geom_rug()
```

Lets look at the reads with multiple hits

```{r}
blast_red_filt %>% 
  group_by(subject) %>% 
  count() %>% 
  filter(n > 1) %>% 
  left_join(blast_red_filt, by = "subject") %>% 
  select(subject, start.subject, end.subject, orientation)
```

```{r}
blast_green_filt %>% 
  group_by(subject) %>% 
  count() %>% 
  ggplot(aes(n)) + 
  geom_histogram(bins = 150, fill="darkgreen") +
  geom_rug()
```

Lets look at the reads with multiple hits

```{r}
blast_green_filt %>% 
  group_by(subject) %>% 
  count() %>% 
  filter(n > 1) %>% 
  left_join(blast_green_filt, by = "subject") %>% 
  select(subject, start.subject, end.subject, orientation)
```

Such reads should not be kept in the analysis

## Filtering, part 2

Filter out:

- reads with multiple green/red regions detected
- reads with only one red or green hit (join first!)
- reads with different orientation of the flanking regions
- reads without full blaSHV genes in between the flanking regions 
or/and -> requires more blasting and subsequent analysis 
- reads with 'wrong' number of repeats between the flanking regions -> divide distance by AU length

### Filter out reads with multiple hits

```{r}
# a function to get IDs of reads with multiple hits of the same 'color'
get_multihits_ids <- function(df){
  # df: filtered blast table
  df %>% 
  group_by(subject) %>% 
  count() %>% 
  filter(n > 1) %>% 
  pull(subject)
}

reads_multiple_hits <- union(get_multihits_ids(blast_red_filt), 
                             get_multihits_ids(blast_green_filt))
```

There are `r length(reads_multiple_hits)` reads with multiple hits

Remove these reads

```{r}
blast_red_filt <- 
  blast_red_filt %>% 
  filter(!subject %in% reads_multiple_hits)

blast_green_filt <-
  blast_green_filt %>% 
  filter(!subject %in% reads_multiple_hits)
```


### Filter out reads with single hits

First, join the tables by subject (read ID)
Then, remove reads with 'lonely' FRs

```{r}
blast_joined <- 
  full_join(blast_red_filt, blast_green_filt, by = "subject")

# reads with only one flanking region should have NA in query.x or query.y
blast_joined %>% filter(is.na(query.x) | is.na(query.y))
```

Remove them

```{r}
blast_joined_filt <-
  blast_joined %>% 
  filter(!is.na(query.x), 
         !is.na(query.y))

blast_joined_filt %>% 
  head()
```

### Filter by orientation

```{r}
# orientation.x and orientation.y should match
blast_joined_filt %>% 
  filter(orientation.x != orientation.y) %>% 
  select(contains('query'), contains('orientation'))
```

There are 18 reads with non-matching orientation of the FRs

Remove them

```{r}
blast_joined_filt <-
  blast_joined_filt %>%
  filter(orientation.x == orientation.y)

nrow(blast_joined_filt)
```


### Filter by number of AUs

using distance between the FRs

```{r}
blast_joined_filt <- blast_joined_filt %>% 
  mutate(green.red.distance = end.subject.x - start.subject.y) 

blast_joined_filt %>% 
  ggplot(aes(green.red.distance)) +
  geom_histogram(bins = 100, fill='steelblue') +
  geom_rug() +
  facet_grid(rows = vars(orientation.x))
```

```{r}
blast_joined_filt %>% 
  mutate(green.red.distance.2 = if_else(green.red.distance < 0, green.red.distance * -1, green.red.distance * 1)) %>%
  ggplot(aes(green.red.distance.2)) +
  geom_histogram(bins = 200, fill='steelblue') +
  geom_rug()
```


Direct:

```{r}
blast_joined_filt %>% 
  select(subject, green.red.distance, orientation.x) %>% 
  filter(orientation.x == "direct") %>% 
  arrange(green.red.distance)
```

Reverse:

```{r}
blast_joined_filt %>% 
  mutate(green.red.distance = end.subject.x - start.subject.y) %>% 
  select(subject, green.red.distance, orientation.x) %>% 
  filter(orientation.x == "reverse") %>% 
  arrange(-green.red.distance)
```

## Length of the reads

```{r}
library(Biostrings)

file_path <- "../results/reads_filtered/replicate_e/CNV_reads_all_3.7k.fasta" 

# Read the file
reads <- readDNAStringSet(file_path)
read_lengths <- width(reads)

reads_len_df <- tibble("subject" = sub("^(.*?) runid=.*", "\\1", names(reads)), 
                       "read.len" = read_lengths)

ggplot(reads_len_df, aes(read.len)) +
  geom_histogram(bins = 300, fill = "steelblue")
```

Join this table with filtered blast results and check the length distribution

```{r}
blast_joined_filt <- left_join(blast_joined_filt,
                               reads_len_df,
                               by = "subject")

blast_joined_filt %>% 
  ggplot(aes(read.len)) +
  geom_histogram(bins = 300)
```

```{r}
blast_joined_filt <-
  blast_joined_filt %>% 
  mutate(green.red.distance.pos = if_else(green.red.distance < 0, green.red.distance * -1, green.red.distance * 1)) 

blast_joined_filt %>% 
  ggplot(aes(green.red.distance.pos, read.len)) +
  geom_point(alpha=0.3) +
  xlab("Distance between FRs") +
  ylab("read length")
```


## BLASTing the blaSHV genes

I can just blast all reads together, then *analyse* the results and join them with FR blast table

```{bash, eval=FALSE}
# requires ca. 23 Gb of RAM
blastn -query resources/genes/blaSHV.fa -subject results/reads_filtered/replicate_e/CNV_reads_all_3.7k.fasta -outfmt 6 -num_alignments 5000000 -num_threads 4 > results/tables/blast_blaSHV.tsv
```

Read the blast results and leave only the reads that have been selected during the previous steps

```{r, message=FALSE, warning=FALSE}
blast_blaSHV <- read_delim("../results/tables/blast_blaSHV.tsv", col_names = F) %>% 
  filter(X2 %in% unique(blast_joined_filt$subject)) %>% 
  filter(X11 > 10*-5)

names(blast_blaSHV) <- c("query", "subject", "identity", "length", "mismatch", 
                        "gaps", "start.query", "end.query", "start.subject", 
                        "end.subject", "e.value", "bit.score") 
blast_blaSHV
```

### Some plots

```{r}
blast_blaSHV %>% 
  ggplot(aes(length)) +
  geom_histogram(bins = 300, fill = "steelblue") +
  geom_rug()
```

```{r}
blast_blaSHV %>% 
  ggplot(aes(identity)) +
  geom_histogram(bins = 300, fill = "steelblue") +
  geom_rug()
```

```{r}
blast_blaSHV %>% 
  ggplot(aes(gaps)) +
  geom_histogram(bins = 50, fill = "steelblue") +
  geom_rug()
```

### How many blaSHV hits per read?

```{r}
blast_blaSHV %>% 
  group_by(subject) %>% 
  count() %>% 
  ggplot(aes(n)) +
  geom_histogram(bins = 100, fill = "steelblue") +
  geom_rug()
```

This histogram corresponds to the previous plot (read length vs distance between FRs) and the distributions of distances between FRs

Maybe blaSHV hits are shorter on the reads with only two balSHV copies?

```{r}
blaSHV_mean_len <- 
  blast_blaSHV %>% 
  group_by(subject) %>% 
  summarise(mean.length = mean(length)) 

blaSHV_counts <- 
  blast_blaSHV %>% 
  group_by(subject) %>% 
  count()

left_join(blaSHV_mean_len, blaSHV_counts, by = "subject") %>% 
  mutate(n = as_factor(n)) %>% 
  ggplot(aes(n, mean.length)) +
  geom_jitter(alpha=0.3, size = 0.5, height = 0.05)
```

Extract the names of the reads with 6 CNV

```{r}
left_join(blaSHV_mean_len, blaSHV_counts, by = "subject") %>% 
  filter(n == 6)
```

use seqkit grep -p %read_id% to extract the read by id

### Count blaSHV genes not hits

sum of blaSHV hits between two FRs divided by blaSHV length (~860 nt)

```{r}
blaSHV_counts <- 
  blast_blaSHV %>% 
  group_by(subject) %>% 
  summarise(sum.bla.hit.len = sum(length)) %>% 
  mutate(n.blaSHV = round(sum.bla.hit.len/860))

blaSHV_counts %>% 
  ggplot(aes(n.blaSHV)) +
  geom_histogram(bins = 200, fill = "steelblue") +
  geom_rug()
```




## blaSHV genes and the distance between FRs

```{r}
# join bla counts and ain blast table containing gree.red.distance
blast_joined_filt <- 
  left_join(blast_joined_filt, 
            blaSHV_counts, 
            by = "subject") %>% 
  mutate(n.blaSHV = replace_na(n.blaSHV, 0))

blast_joined_filt %>% 
  ggplot(aes(n.blaSHV, green.red.distance.pos - (2*1677+820))) +
  geom_point(alpha=0.3, color="steelblue") +
  geom_smooth(method = "lm") # use your custom function dist = X * n.blaSHV
  

```






