---
title: "BL-CNV analysis"
author: "AG"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    df_print: paged
    code_folding: hide
    fig_width: 10
    fig_height: 6
    theme: cerulean
    highlight: kate
    toc_float: yes
  pdf_document:
    
    toc: yes
    toc_depth: 2
editor_options:
  markdown:
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

# Test data set

We have a small data set to test the approach to counting BL copy number

## Map reads

Using *minimap2*

```{bash mapping, eval=FALSE}
# reads with 3 AU onto the plasmid with 3 AU
minimap2 -ax map-ont reads/DA61218_ref_3AU.fasta data/DA61218_filt.fq > mapping/DA61218_3AU.sam

# reads with 1 AU onto the same plasmid with 1 AU
minimap2 -ax map-ont reads/DA61218_ref.fasta reads/DA61218_filt.fq > mapping/DA61218.sam
```


## Get coverage

Using *samtools* and *bedtools*

```{bash bedtools, eval=FALSE}
# same thing for both plasmids

# convert sam to bam
samtools view -b mapping/DA61218_3AU.sam | samtools sort -o mapping/DA61218_3AU.bam -O BAM -@ 2 && samtools index mapping/DA61218_3AU.bam && samtools coverage -m mapping/DA61218_3AU.bam

# get coverage histogram
bedtools coverage -a bedfiles/DA61218_3AU.bed -b results/DA61218_3AU.bam -d > tables/DA61218_3AU_cov.tab
```


## Read and combine coverage data

```{r coverage, warning=FALSE, message=FALSE, echo=FALSE}
# three BLA
cov3 <- read_delim("../results/tables/DA61218_3AU_cov.tab", delim = "\t", 
                   escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
cov3$plasmid <- "3 AU"
cov3 <- cov3 %>% select(X4, X5, plasmid)

# one BLA
cov <- read_delim("../results/tables/DA61218_cov.tab", delim = "\t", 
                  escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
cov$plasmid <- "1 AU"
cov <- cov %>% select(X4, X5, plasmid)

# bind them
cov_all <- bind_rows(cov, cov3)

cov_all
```

## Plot coverage

### Coverage plot of the same set of reads (3 AU) on to plasmids with 1 AU and 3 AU.


```{r plot, echo=FALSE, fig.width=10}
ggplot(cov_all, aes(X4, X5))+ 
  geom_point(size=0.2, alpha=0.2, aes(color=plasmid)) + 
  ggtitle("coverage") +
  xlab("pos") +
  ylab("cov")
```

### Coverage plot of the 1 AU plasmid between positions 25k - 30k

```{r, echo=FALSE}
ggplot(cov %>% filter(X4 > 25100, X4 < 29400), aes(X4, X5)) + 
  geom_point(size=0.2, alpha=0.2) + 
  ggtitle("coverage") +
  xlab("pos") +
  ylab("cov")
```



## Lowest coverage

**One copy**

```{r, echo=FALSE}
cov %>% arrange(X5) %>% head()
```

**Three copies**

```{r, echo=FALSE}
cov3 %>% arrange(X5) %>% head()
```

## Copy number of BL genes

I should calculate it for every read containing both Flanking Regions (FR), length 300 bp.

FR1: 24862 - 25162 (5'-end of BL gene in reference)

FR2: 29388 - 29688 (3'-end of BL gene in reference)

How to find both FR in a read?

```{bash, eval=FALSE}
# filter out too short reads
filtlong --min_length 3500 data/DA61218.fq.gz > data/DA61218_filt.fq

# convert fastq to fasta
seqkit fq2fa data/DA61218_filt.fq > data/DA61218_filt.fasta

# RUN BLAST
#
# -sorthsps <Integer, (>=0 and =<4)>
#   Sorting option for hps:
#     0 = Sort by hsp evalue,
#     1 = Sort by hsp score,
#     2 = Sort by hsp query start,
#     3 = Sort by hsp percent identity,
#     4 = Sort by hsp subject start
#   Not applicable for outfmt != 0
#
# -num_alignments <Integer, >=0>
#   Number of database sequences to show alignments for
#   Default = `250'
#    * Incompatible with:  max_target_seqs


blastn -query reads/flanking_region_1.fasta -subject reads/DA61218_filt.fasta -outfmt 6 -num_alignments 6000 > tables/blast_fr1.tsv

blastn -query reads/flanking_region_2.fasta -subject reads/DA61218_filt.fasta -outfmt 6 -num_alignments 6000 > tables/blast_fr2.tsv
```


### FR1

Identity and Length Statistics

```{r, message=FALSE, echo=FALSE}
fr1 <- read_delim("../results/tables/blast_fr1.tsv", col_names = F, col_select = c(X1, X2, X3, X4, X9, X10)) 

names(fr1) <- c("query", "subject", "identity", "length", "start.subject", "end.subject")

fr1 %>% select(identity, length) %>% summary()
```

Identity and hit length are pretty low sometimes

Filter out hits shorter than ~ 250 bp

```{r, echo=FALSE}
fr1 <- filter(fr1, length > 250)

fr1 %>% select(identity, length) %>% summary()
```

Now number of hits per read is 1

```{r, echo=FALSE}
fr1 %>% group_by(subject) %>% summarise(N=n()) %>% arrange(-N)
```

I should keep only reads having both hits on *the same* strand!

```{r, echo=FALSE}
fr1 <- fr1 %>% mutate(pos_strain = (end.subject - start.subject) > 0, .keep="all") 

fr1
```



### FR2

Identity and length statistics

```{r, message=FALSE, echo=FALSE}
fr2 <- read_delim("../results/tables/blast_fr2.tsv", col_names = F, col_select = c(X1, X2, X3, X4, X9, X10)) 

names(fr2) <- c("query", "subject", "identity", "length", "start.subject", "end.subject")

fr2 %>% select(identity, length) %>% summary()
```

Number of hits per read

```{r, echo=FALSE}
fr2 %>% group_by(subject) %>% summarise(N=n()) %>% arrange(-N)
```

In each table I've got 1 hit per read. That's what I should see.

Remove hits on negative strains

```{r, echo=FALSE}
fr2 <- fr2 %>% mutate(pos_strain = (end.subject - start.subject) > 0, .keep="all") 

fr2
```

Now I need to find such positive strand rows that have *both* FR and negative strand rows that have both FR,
then I can join them together

### Only positive strand

```{r}
fr1_pos <- fr1 %>% filter(pos_strain)
fr2_pos <- fr2 %>% filter(pos_strain)

fr12_pos <- inner_join(fr1_pos, fr2_pos, by="subject")
```


### Only negative strand

```{r}
fr1_neg <- fr1 %>% filter(!pos_strain)
fr2_neg <- fr2 %>% filter(!pos_strain)

fr12_neg <- inner_join(fr1_neg, fr2_neg, by="subject")
```

### Joining both data sets

```{r, echo=FALSE}
fr12 <- bind_rows(fr12_pos, fr12_neg)

fr12
```

### Copy number distribution

find distances between FRs and divide by 3500

```{r, echo=FALSE}
n_copies <- fr12 %>% 
  transmute(n.cop = round(abs(start.subject.y - end.subject.x) / 3500)) 

ggplot(n_copies, aes(n.cop)) + 
  geom_histogram(binwidth = 0.5, fill="steelblue") + 
  xlab("AU") +
  ggtitle("BL-genes copy number distribution", subtitle = paste0("N total spanning reads ", nrow(n_copies)))

```

Number of spanning reads just for one barcode is ~ 190

# Big data set

I've got a much bigger data set,

I have made a pipeline that run the above analysis on this data set

## BLAST tables

Let's have a look at the blast tables: identity, lengths

```{r, message=FALSE}
fr1 <- read_delim("../results/tables/FAU50052_FR1_blast.tsv", col_names = F, col_select = c(X1, X2, X3, X4, X9, X10)) 

names(fr1) <- c("query", "subject", "identity", "length", "start.subject", "end.subject")

fr1$region <- "fr1"

fr2 <- read_delim("../results/tables/FAU50052_FR2_blast.tsv", col_names = F, col_select = c(X1, X2, X3, X4, X9, X10)) 

names(fr2) <- c("query", "subject", "identity", "length", "start.subject", "end.subject")

fr2$region <- "fr2"

blast.fr <- bind_rows(fr1, fr2)
```

### Raw

```{r}
blast.fr %>% group_by(region) %>% 
  count()
```


#### Identity

```{r}
ggplot(blast.fr, aes(identity)) +
  geom_histogram(fill = "steelblue", bins = 100 ) +
  geom_rug() +
  coord_trans(y = "sqrt") +
  facet_grid(rows = vars(region))
```

#### Length

```{r}
ggplot(blast.fr, aes(length)) +
  geom_histogram(fill = "steelblue", bins = 100 ) +
  geom_rug() +
  coord_trans(y = "sqrt") +
  facet_grid(rows = vars(region))
```
#### Identity v Length

```{r}
ggplot(blast.fr, aes(length, identity)) +
  geom_point(color = "steelblue", alpha = 0.2, size = 0.4) +
  facet_grid(rows = vars(region))
```


### After filtering 

min len 250

```{r}
blast.fr %>% filter(length >= 250) %>% 
  group_by(region) %>% 
  count()
```


#### Identity

```{r}
ggplot(blast.fr %>% filter(length >= 250), aes(identity)) +
  geom_histogram(fill = "steelblue", bins = 100 ) +
  geom_rug() +
  coord_trans(y = "sqrt") +
  facet_grid(rows = vars(region))
```

#### Length

```{r}
ggplot(blast.fr %>% filter(length >= 250), aes(length)) +
  geom_histogram(fill = "steelblue", bins = 100 ) +
  geom_rug() +
  coord_trans(y = "sqrt") +
  facet_grid(rows = vars(region))
```

#### Identity v Length

```{r}
ggplot(blast.fr %>% filter(length >= 250), aes(length, identity)) +
  geom_point(color = "steelblue", alpha = 0.2, size = 0.4) +
  facet_grid(rows = vars(region))
```

## CNV

```{r}
parse_blast_df <- function(df, filter_on_positive=TRUE){
  # add column indicating that the hit is on positive strain or not
  df <- df %>% mutate(pos_strand = (end.subject - start.subject) > 0, .keep = "all") 
  # find such positive strand rows that have both FR, and negative strand rows that have both FR
  if (filter_on_positive){
    df_strand <- df %>% filter(pos_strand)
  } else {
    df_strand <- df %>% filter(!pos_strand)
  }
  return(df_strand)
}

# a function to join processed BLAST output
join_filtered_blast <- function(df1, df2, on_pos){
  parsed_list <- lapply(c(df1, df2), function(x){
    parse_blast_df(x, filter_on_positive=on_pos)
    })
  df12_joined <- inner_join(parsed_list[[1]], parsed_list[[2]], by = "subject")
  return(df12_joined)
}

# COLLECT POS STRAND ROWS W BOTH FR
fr1.filt <- fr1 %>% filter(length >= 250)
fr2.filt <- fr2 %>% filter(length >= 250)

fr1_pos <- parse_blast_df(fr1.filt, filter_on_positive = T)
fr2_pos <- parse_blast_df(fr2.filt, filter_on_positive = T)
fr12_pos <- inner_join(fr1_pos, fr2_pos, by = "subject")

# COLLECT NEG STRAND ROWS W BOTH FR
fr1_neg <- parse_blast_df(fr1.filt, filter_on_positive = F)
fr2_neg <- parse_blast_df(fr2.filt, filter_on_positive = F)
fr12_neg <- inner_join(fr1_neg, fr2_neg, by = "subject")

# BIND ROWS
fr12_all <- bind_rows(fr12_pos, fr12_neg)

# FIND CNV
n_copies <- fr12_all %>% 
  transmute(n.copies = round(abs(start.subject.y - end.subject.x) / 3500)) 

# PLOT
ggplot(n_copies, aes(n.copies)) + 
  geom_histogram(binwidth = 0.5, fill = "steelblue") + 
  coord_trans(y = "sqrt") +
  xlab("AU") +
  ggtitle("BL-genes copy number distribution", subtitle = paste0("N total spanning reads ", nrow(n_copies)))
```

```{r}
n_copies %>% group_by(n.copies) %>% summarize(counts = n())
#fr12_all %>% transmute(n.cop = abs((start.subject.y - end.subject.x) / 3500)) %>% filter(n.cop > 0.2, n.cop < 0.5) %>% arrange(n.cop)
```


```{r}
x <- n_copies %>% group_by(n.copies) %>% summarize(counts = n())
write.csv(x, "../results/CNV/FAU50052_cnv_counts.csv", row.names = F, quote = F)
```

